components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt

```diff
@@ -0,0 +1,173 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.logins
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.storage.Login
+import mozilla.components.concept.storage.LoginStorageDelegate
+import mozilla.components.lib.dataprotect.SecureAbove22Preferences
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * A type of persistence operation, either 'create' or 'update'.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+enum class Operation { CREATE, UPDATE }
+
+internal const val PASSWORDS_KEY = "passwords"
+
+/**
+ * [LoginStorageDelegate] implementation.
+ *
+ * An abstraction that handles the persistence and retrieval of [LoginEntry]s so that Gecko doesn't
+ * have to.
+ *
+ * In order to use this class, attach it to the active [GeckoRuntime] as its `loginStorageDelegate`.
+ * It is not designed to work with other engines.
+ *
+ * This class is part of a complex flow integrating Gecko and Application Services code, which is
+ * described here:
+ *
+ * - GV finds something on a page that it believes could be autofilled
+ * - GV calls [onLoginFetch]
+ *   - We retrieve [Login]s from all matching domains (if any) from [loginStorage]
+ *   - We return these [Login]s to GV
+ * - User submits their credentials
+ * - GV decides to save the login
```

Comment:
"decides" here means that user is presented with a prompt, and selects to save/update the login (based on if it's a dupe or not), right? Can you elaborate here? This would also help tie the two pieces of the puzzle together for the future readers.

Reply:
I'll flesh out the comment flow a bit.  This bullet doesn't involve any UI, presumably it's GV seeing that something that looks like a submit button was clicked after something that looks like username/password fields were filled, but the actual logic is a black box from our perspective.

I'll add a bullet point for our prompt as well, to help clarify.